### PR TITLE
Addon system

### PR DESCRIPTION
**Pull request by Globegitter**
This is nowhere near to be merged.

Please add any feedback below and this list is open to be updated. 

Goals for this addon system/PR:

- [x] Addon discover for `sane generate` having a default generator based on the addon-name
- [ ] Support for multiple generators in an addon
- [x] Add `sane install` command that installs an and addon runs the default generator
- [ ] Add support for node-machine (http://node-machine.org/) structure as addons
    - [x] Basic support that node-machines are being recognized and called
- [ ] Add support for custom generators:
   - [x] To add custom templates to the backend and frontend
       - [x] Support file diff if the file already exists
      - [ ] See how ember-cli-admin are supporting pod structure: https://github.com/ember-admin/ember-cli-admin/tree/master/blueprints/table-view-resource/files/app as well as https://github.com/ember-admin/ember-cli-admin/blob/master/blueprints/table-view-resource/index.js
   - [x] To install npm packages for the backend (including sails hooks)
   - [x] To install ember addons on the frontend
   - [ ] To install bower packages on the frontend
   - [ ] To install npm packages on the frontend
   - [ ] To add to the ember-cli config file
   - [ ] To add to sails config files?
   - [ ] To run `ember generate`
   - [ ] or `sails generate` commands for creating files dynamically*
- [ ] Keep track if a generator has already been run (e.g. through the install command), within the .sane-cli
- [ ] Support custom folder names for client/server
- [ ] write a `sane generate addon` or `sane new addon` or `sane addon` command which scaffolds out a skeleton template
- [ ] Add 'dryRun' option to generators so they don't actually `npm install` but just add the packages to the `package.json` for better testability (part of the reason for the failing/flakey tests)
- [ ] Add uninstall command? (probably not in this iteration)

Any other ideas @kriswill @mgenev ? Especially, what do you think about the scaffolding command? Should we re-use one of the existing ones or add a new command?

Also @IanVS @albertosouza @mphasize @huafa @DiknT @bsocius what do you think? Would love some more input.

*This could easily support user-input so files can be name dynamically. So for example for sane-auth right now the model is called `user` on the frontend and backend, but that could be made dynamically by passing `sane install sane-auth customer` or something like that and then the model is called `customer`.

### CODE REVIEW

**Comment by Globegitter** [comment]
So, right now the system is extremely simple and expects the default folder-names (client/server) to function properly. So you run `sane install <addon-name>` which will automatically call `sane genrate <addon-name>` once installed.

The generator on the other hand detects any addon that has the `sane-addon` keyword as specified here https://github.com/Globegitter/sane-auth/blob/master/package.json#L12. Furthermore `machinepack` needs to be defined: https://github.com/Globegitter/sane-auth/blob/master/package.json#L26 as per http://node-machine.org/spec/machinepack]

Based on this addon I would also like to see what other people think about using machines.

Right now all the code and templates have to lie within the `generate` folder and no sub-folders/multiple generators are supported, but that will most likely change soon. 

Just have to see how that can be done best using machinepacks.

Also another thought that came up was making the addon system within sane-cli a machine itself that just talks to other machines. Have to investigate that.

On another note, fixing windows tests shouldn't be that hard anymore since the efforts of @IanVS :)

**Comment by Globegitter** [comment]
Just jotting down some more thoughts about the addon system itself being a machine:

Right now the addons are a machine and while that has certain benefits I also think it is a bit like stretching the system for its original purposes (e.g. see that sane-auth basically has no code, it just returns an exits.success). It is still useful to have a description of what exactly each file does and also to define what you are using in the exits.success.

But what now that I am more thinking about it would make a lot of sense as well, the addon system itself to be a machine (that possibly loads other machines): This basically means that within the addon-system the input expectations for every addon is defined (so for the addon `inputs` we could define `addEmberAddons` , `addToConfig`, etc) and get easier/nicer error message when an addon uses that wrongly.

**Comment by mgenev** [comment]
I reviewed everything so far, except actually, the way it behaves as a machine so I can't comment on that, but the rest seems very logical to me - great work. I do agree that certain customization options need to be added and perhaps some abstraction will be needed to avoid bloat in the generate command file, but other than that so far so good.

**Comment by Globegitter** [comment]
For discussion about addon naming conventions and generator name-spacing, see https://github.com/Globegitter/sane-auth/issues/8

**Comment by Globegitter** [comment]
How can we easily update addons? Think about that. Looking at the sane-auth addon there are a few different main components:
 - [ ] the generated template files
 - [ ] the dependencies added to the project
 - [ ] the addition to the ember config files (which brings the question, do we want similar addition to sails configs)
- [ ] the commands it runs, such as `ember g resource` etc.

Have to identify what currently is easy to upgrade and what not, plus if we could improve anything about that.

**Comment by mgenev** [comment]
About "the generated template files", Ember solves the problem by keeping them in a node package and importing them at build time and resolving them at runtime by the container. This sounds quite logical to me, on the Ember side we can easily piggy back, by making our sane-addon install an ember-cli-addon for the feature. 
On the Sails side I have no solution, it may be worth talking to the Sails team.

**Comment by kriswill** [comment]
@mgenev yeah, I agree.  Any template generated files have to be somehow closed, with extension points made available for customization.  Example:  sane-auth generates controllers, models, policy, config, etc.  If the files for these things were in a hook, node-machine, etc., then they would be essentially frozen, making upgrading them easier, since they will not have been modified by the user.  However, we still need to provide the ability to customize behaviour, etc.

**Comment by mgenev** [comment]
About dependencies, aren't most dependencies easily added by  running a  npm install --save and bower? 

About configs, yes I imagine we need to touch the sails configs as well.

**Comment by Globegitter** [comment]
Now that I switched to babel, I need to make sure it does not skip to transpile the locally installed sane-cli. It does so by default: babeljs.io/docs/usage/require/

**Comment by Globegitter** [comment]
@IanVS for some reason your helper unit test is failing with `TypeError: object is not a function`. Any idea why?

**Comment by IanVS** [comment]
The helper basically returns a string (the path to the ember-cli binary).  That was causing trouble with rewire, and perhaps babel doesn't like it either?  Is that an antipatern in Node?

**Comment by Globegitter** [comment]
@IanVS Not that I am aware, especially because `returns local ember by default` does work. I also updated chai to `2.x` so maybe that changed something about the `expect().to.be.ok()` syntax?

I also updated mocha to `2.1.x` but don't expect that to be any issue.

**Comment by IanVS** [comment]
OK yes, chai seems to have changed.  I can't do a full PR, but try this (remove the function call at the end):

```
expect(ember).to.be.ok;
```

but then you need to add this to the top of the file to make ESLint happy:

```
/*eslint no-unused-expressions:0 */
```

**Comment by Globegitter** [comment]
Thank you @IanVS . Now we just have to fix windows tests and we are good to go!

**Comment by IanVS** [comment]
It might be because your 'only' option for babel is using escaped forward slashes, and Windows uses backslashes. I think you shouldn't escape the forward slashes. See: https://www.npmjs.com/package/glob#windows

**Comment by Globegitter** [comment]
@IanVS the tests are looking much better now. It does get quite far. The log shows `sane-auth succesfully installed. Running it's default generator now...`. And the generator also detects `module found: sane-auth`.

Not sure why it just stops then. Probably best to test on a windows machine directly and add some more debugging statements.

File: lib/tasks/generate/runMachine.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    var promise = new Promise(function (resolve) { //reject
+      blueprint.client.exec({
+        success: async function (actions) {
+          //only install ember-addons addons if --skip-npm is not set
+          if (!options.skipNpm) {
+            for (var emberAddon of actions.addEmberAddons) {
+              var addonSemVer = emberAddon.name + '@' + emberAddon.target;
+              var emberArgs = ['install', addonSemVer];
+              try {
+                await spawn('ember', emberArgs, { stdio: 'inherit', cwd: path.join(projectRoot, clientName) });
```
Comment (IanVS) [inline lib/tasks/generate/runMachine.js]: Need to use ember helper here, not hardcoded `'ember'`.  Again on line 40 as well.

Comment (Globegitter) [inline lib/tasks/generate/runMachine.js]: Ahh thank you!

**Comment by Globegitter** [comment]
Yes! @IanVS we got passing tests! But lets see if traceur is any faster than babel.

Babel results:
```js
travis:
Node 0.12 - 2m49s -  11 passing (52s)
  sane generate
    resource/api user (13671ms)
    resource/api user name:string age:number (8860ms)
  sane install
    sane-auth and runs generator (11407ms)
  sane help
    displays commands, it's aliases and the correct cli version (1221ms)

io.js - 2m57s - 11 passing (54s)
  sane generate
    resource/api user (12233ms)
    resource/api user name:string age:number (8388ms)
  sane install
    sane-auth and runs generator (11686ms)
  sane help
    displays commands, it's aliases and the correct cli version (1265ms)

appveyor:
21min20s, with tests themselves passing within 5min.
  sane generate
    resource/api user (67379ms)
    resource/api user name:string age:number (50874ms)
  sane install
    sane-auth and runs generator (55553ms)
  sane help
    displays commands, it's aliases and the correct cli version (6455ms)
```

**Comment by Globegitter** [comment]
@IanVS I meant babel vs traceur. Seems there is some difference: io.js is 11s faster, node.js 2s and windows roughly 1min.

Reason I wanted to check is because babel always just seemed to have the tendency to be a bit slower compared to traceur in my tests. Also just read: https://github.com/babel/babel/issues/476#issuecomment-93969324

Now is the question, improved development experience vs performance :P

**Comment by IanVS** [comment]
Well, as @jamemackson mentioned in another thread, creating a new project only happens once per project. 

Could also keep it as an option of switching back if people start to complain that it takes too long.

**Comment by IanVS** [comment]
I think that the way babel is being used here is causing some problems.  Babel recommends to compile before distributing a package on npm (see: https://github.com/babel/babel/pull/1026#issuecomment-80946352).  Is that an option?

Otherwise, it does seem like `ignore: false` is necessary after all.  I'll try to play around with compiling and see how that would look.

**Comment by Globegitter** [comment]
@IanVS awesome this is looking good now!

**Comment by Globegitter** [comment]
Ok, fixes left:
 - [ ] the last `npm install` hangs on `sane new` but does install correctly. For some reason it does not seem to notice when the process is done. Investigate, or think about switching to `enpeem`
- [ ] Fix `sane-auth`

**Comment by IanVS** [comment]
@Globegitter what needs to be done to fix `sane-auth`?

**Comment by mgenev** [comment]
Ember-simple-auth changed the interface, still need to look into it

**Comment by Globegitter** [comment]
@IanVS seems there is nothing we can do directly: https://github.com/simplabs/ember-cli-simple-auth/issues/26

But have to test `sane's` behaviour. It should probably still finish the complete installation or at least provide a `--force` flag to finish.

**Comment by IanVS** [comment]
Do you mean add an acceptance test?  The commit I added fixes it when I test on my machine.  The `'--verbose'` option was causing problems if `stdio: "inherit"` was not used in the spawn.  So I removed the verbose and it works correctly (finishes the installation).

**Comment by Globegitter** [comment]
Sorry, wasn't clear at all. I don't actually know how sane handles the error we currently have with `sane-auth/ember-simple-auth`. So I mean I need to test to see the behaviour Martin described. There might not be that much to do.

**Comment by IanVS** [comment]
Ah, got it, yes.

**Comment by Globegitter** [comment]
LGTM.

**Pull request merged**